pkg/blobcache/blobcache.go

```diff
@@ -0,0 +1,410 @@
+		f.Close()
+	}
+	return nil
+}
+
+func (r *blobCacheReference) NewImage(ctx context.Context, sys *types.SystemContext) (types.ImageCloser, error) {
+	src, err := r.NewImageSource(ctx, sys)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating new image %q", transports.ImageName(r.reference))
+	}
+	return image.FromSource(ctx, sys, src)
+}
+
+func (r *blobCacheReference) NewImageSource(ctx context.Context, sys *types.SystemContext) (types.ImageSource, error) {
+	src, err := r.reference.NewImageSource(ctx, sys)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating new image source %q", transports.ImageName(r.reference))
+	}
+	return &blobCacheSource{reference: r, source: src}, nil
+}
+
+func (r *blobCacheReference) NewImageDestination(ctx context.Context, sys *types.SystemContext) (types.ImageDestination, error) {
+	dest, err := r.reference.NewImageDestination(ctx, sys)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating new image destination %q", transports.ImageName(r.reference))
+	}
+	return &blobCacheDestination{reference: r, destination: dest, applied: make(map[digest.Digest]struct{})}, nil
+}
+
+func (s *blobCacheSource) Reference() types.ImageReference {
+	return s.reference
+}
+
+func (s *blobCacheSource) Close() error {
+	return s.source.Close()
+}
+
+func (s *blobCacheSource) GetManifest(ctx context.Context, instanceDigest *digest.Digest) ([]byte, string, error) {
+	return s.source.GetManifest(ctx, instanceDigest)
+}
+
+func (s *blobCacheSource) getBlob(ctx context.Context, blobinfo types.BlobInfo) (bool, io.ReadCloser, int64, error) {
+	present, size, err := s.reference.HasBlob(blobinfo)
+	if err != nil {
+		return false, nil, -1, err
+	}
+	if !present {
+		rc, size, err := s.source.GetBlob(ctx, blobinfo)
+		if err != nil {
+			return false, nil, -1, errors.Wrapf(err, "error reading blob from source image %q", transports.ImageName(s.reference))
+		}
+		return false, rc, size, nil
+	}
+	for _, directory := range s.reference.directories {
+		if directory == "" {
+			continue
+		}
+		for _, isConfig := range []bool{false, true} {
+			filename := filepath.Join(directory, makeFilename(blobinfo.Digest, isConfig))
+			f, err := os.Open(filename)
+			if err == nil {
+				return true, f, size, nil
+			}
+		}
+	}
+	rc, size, err := s.source.GetBlob(ctx, blobinfo)
+	if err != nil {
+		return false, rc, size, errors.Wrapf(err, "error reading blob from source image %q", transports.ImageName(s.reference))
+	}
+	return false, rc, size, nil
+}
+
+func (s *blobCacheSource) GetBlob(ctx context.Context, blobinfo types.BlobInfo) (io.ReadCloser, int64, error) {
+	_, rc, size, err := s.getBlob(ctx, blobinfo)
+	return rc, size, err
+}
+
+func (s *blobCacheSource) GetSignatures(ctx context.Context, instanceDigest *digest.Digest) ([][]byte, error) {
+	return s.source.GetSignatures(ctx, instanceDigest)
+}
+
+func (s *blobCacheSource) LayerInfosForCopy(ctx context.Context) ([]types.BlobInfo, error) {
+	return nil, nil
```

Comment:
An image that we committed to `containers-storage` will generally use the uncompressed diffIDs in its manifest, since at commit-time, we don't know that we'll be able to supply layer blobs in a compressed form that matches a given digest, so the underlying `LayerInfosForCopy()` doesn't return values that match the blobs that we've cached.
In order to force a copying operation to use the blobs from the cache, we also have to "break" `Commit()` to have it record the original compressed digest values when it creates a manifest, which means that copying the resulting image can fail unless the cache is also there to supply blobs at copy-time.

Reply:
> An image that we committed to `containers-storage` will generally use the uncompressed diffIDs in its manifest, since at commit-time, we don't know that we'll be able to supply layer blobs in a compressed form that matches a given digest

AFAICS it’s the opposite: `DesiredLayerCompression`+`PutBlob` pretend to accept the unmodified compressed blobs, and the original manifest referring to compressed blobs is recorded — as well as returned by `GetManifest`.

> so the underlying `LayerInfosForCopy()` doesn't return values that match the blobs that we've cached.

Ah, right — `LayerInfosForCopy` is where the uncompressed blob digests get substituted, and we don’t want that happening here, at least _not always_.

> In order to force a copying operation to use the blobs from the cache, we also have to "break" `Commit()` to have it record the original compressed digest values when it creates a manifest, which means that copying the resulting image can fail unless the cache is also there to supply blobs at copy-time.

AFAICS this implementation forwards to the underlying `Commit` without modification, so I’m not sure what this  is referring to.

---

Overall, though, the choice to blindly use unmodified `LayerInfosForCopy` here seems rather radical to me — now the cached blobs _must_ be present or copying from `containers-storage:` will fail.  That looks brittle and very non-cache-like (in particular, the only time the cache can be enabled is when the local image storage is empty).

(Arguably the outlier here is `containers-storage:` pretending to store the compressed blobs, not the cache. I’m not really strongly arguing for changing that at this point, though.)